### PR TITLE
Rendered html fixes and tuning

### PR DESCRIPTION
**Pull request by ellisonbg**
This affects:

* Markdown cells
* Markdown/HTML output
* Markdown rendered documents.

Bring it up to the level of the classic notebook...

### CODE REVIEW

**Comment by ellisonbg** [comment]
Fixes #697

**Comment by sccolbert** [comment]
`.jp-mod-whatever` classes should always be used in conjunction with the thing they are modifying. i.e. `.jp-SomeWidget.jp-mod-rendered`, never simply on their own. This is a semantic organizational thing. The `mod` means "modifier" and you want to know *what* it's modifying. 

If `.jp-mod-renderedHTML` is not actually a modifier (which it doesn't seem like it is), it should be renamed.

The class name used before was `.jp-Rendered-html`, which (if JLab is using Phosphor's convention) is a class name which denotes the `html` subcontrol of a `Rendered` widget.

**Comment by sccolbert** [comment]
It looks like all of the class names in this file are bonkers:
https://github.com/ellisonbg/jupyterlab/blob/f4cc958790f1ed6b22c8eb2301d14dc9c72333be/src/renderers/index.ts

**Comment by sccolbert** [comment]
There is no way to distinguish a latex widget from a markdown widget because they are all given the same CSS names. That is wrong.

**Comment by sccolbert** [comment]
cc @blink1073 @afshin

**Comment by jasongrout** [comment]
Are the phosphor CSS naming conventions documented somewhere? I've been (blissfully) ignorant about CSS naming up to this point.

**Comment by sccolbert** [comment]
Unfortunately, no. But they absolutely should be. That's why I cc'ed Darian, cause he and I hashed this out over several conversations. We can do the same and make sure we've picked something that suits everyone.

**Comment by ellisonbg** [comment]
Bonkers, don't think that is one of our goals for JupyterLab ;-)

I think those CSS class names have been there a while. @blink1073 walked me through some of the CSS naming stuff today so I can try to clean this up. My understanding is this:

* `jp-ClassName` for each class.
* `jp-ClassName-child` for all children
* `jp-mod-something` for classes that modify state.

Seems like `jp-Rendered` should actually be `jp-mod-rendered`?

And I will create a CSS class for each renderer.

**Comment by sccolbert** [comment]
Without knowing the full details of the use case, that sounds more like it.

If you want to discuss, I'll hang out on appear.in/phosphor for a few minutes. If not, we can talk tomorrow.

It seems like I missed the Gitter discussion on this today, and Steven didn't have the fully correct mental model. This is my fault. Sorry!

**Comment by ellisonbg** [comment]
The only subtle point is that the `IRenderer.render` API returns widgets, some of which are raw `Widget` instances. So the relevant class names in this case are not the names of the renderer classes (which are not Widget subclasses), but rather the widget's returned by the render method.

Some of those are named classes (`HTMLWidget`, `MarkdownWidget`) but others are raw `Widget` instances. I propose we use the following naming convention for all of those:

* `jp-RenderedHTML`
* `jp-RenderedLatex`

And rename the named classe to match this convention.

Will try this after dinner...no more bonkers...

**Comment by ellisonbg** [comment]
@sccolbert OK I think I have a pretty good refactor of all this done now. Still need to finish up some of the CSS work, but the TypeScript stuff is ready for review.

File: src/renderers/widget.ts [hunk 1]
```diff
@@ -0,0 +1,350 @@
+// Copyright (c) Jupyter Development Team.
```
Comment (sccolbert) [inline src/renderers/widget.ts]: I might name this file `widgets.ts` instead of the singular form. /pedantry

File: src/renderers/widget.ts [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+ */
+export
+class RenderedMarkdown extends Widget {
+  /**
+   * Construct a new markdown widget.
+   */
+  constructor(options: RenderMime.IRenderOptions) {
+    super();
+    this.addClass(MARKDOWN_CLASS);
+    this.addClass(HTML_CLASS);
```
Comment (sccolbert) [inline src/renderers/widget.ts]: I don't agree with this. `RenderedMarkdown` does not inherit from `RenderedHTML`, so it shouldn't have that CSS class. Maybe we need a token base class which does nothing except add a common CSS class, which is inherited by every widget in this file..

File: src/renderers/widget.ts [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (sccolbert) [inline src/renderers/widget.ts]: Too many extra lines at EOF

**Comment by sccolbert** [comment]
TS looks mostly good. I made a few inline comments.

File: src/renderers/widget.ts (hunk 1 above)
Comment (jasongrout) [inline src/renderers/widget.ts]: The name was to be consistent with other plugins, but I see your point.

**Comment by ellisonbg** [comment]
@sccolbert want to review this?

File: src/renderers/base.css [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+
+.jp-RenderedHTMLCommon {
+  color: black;
+  font-size: 14px;
+  line-height: 20px;
+}
+
+.jp-RenderedHTMLCommon em {
```
Comment (blink1073) [inline src/renderers/base.css]: We've been using two spaces between blocks.

Comment (blink1073) [inline src/renderers/base.css]: IIRC I had copy-pasted this and replaced the class name earlier (my bad).

File: src/renderers/widget.ts [hunk 5]
```diff
@@ -0,0 +1,354 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+
```
Comment (blink1073) [inline src/renderers/widget.ts]: Extra blank line.

File: src/renderers/widget.ts [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+    }).catch(err => {
+      console.log(`No CodeMirror mode: ${lang}`);
+      console.log(`Require CodeMirror mode error: ${err}`);
+      callback(null, code);
+    });
+  }
+});
+
+
+/*
```
Comment (blink1073) [inline src/renderers/widget.ts]: Ill-formatted start and end of quote.

File: src/renderers/widget.ts [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+ * A widget for displaying any widget whoes representation is rendered HTML
+ * */
+export
+class RenderedHTMLCommon extends Widget {
+  /* Construct a new rendered HTML common widget.*/
```
Comment (blink1073) [inline src/renderers/widget.ts]: All of our comments are multi-line unless within a code block.

File: src/renderers/widget.ts [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+  /**
+   * A message handler invoked on an `'after-attach'` message.
+   */
+  onAfterAttach(msg: Message): void {
+    typeset(this.node);
+  }
+}
+
+
+export
```
Comment (blink1073) [inline src/renderers/widget.ts]: Missing class and constructor docs from here down.

File: src/renderers/widget.ts [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+/**
+ * A widget for displaying LaTeX output.
+ */
+export
+class RenderedLatex extends Widget {
+  /**
+   * Construct a new latex widget.
+   */
+  constructor(options: RenderMime.IRenderOptions) {
+    super();
```
Comment (blink1073) [inline src/renderers/widget.ts]: Need to pass options to base class.

File: src/renderers/widget.ts [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+    this.node.textContent = options.source;
+    this.addClass(LATEX_CLASS);
```
Comment (blink1073) [inline src/renderers/widget.ts]: We've been adding the semantic class right after `super()` by convention.

File: src/renderers/widget.ts [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+    this.addClass(TEXT_CLASS);
+  }
+}
+
+
+export
+class RenderedJavascript extends Widget {
+
+  constructor(options: RenderMime.IRenderOptions) {
+    super();
```
Comment (blink1073) [inline src/renderers/widget.ts]: idem

**Comment by blink1073** [comment]
I concur overall, but this has a merge conflict after #703.

**Comment by ellisonbg** [comment]
Rebased, should be ready to go...

**Comment by ellisonbg** [comment]
Also, I am going to start on the CSS proprocessor stuff now...

**Comment by ellisonbg** [comment]
Don't merge this until #717 is in.

**Comment by ellisonbg** [comment]
Replaced by #718

**Pull request closed without merging**